app/renderer/components/navigation/navigator.js

```diff
@@ -0,0 +1,342 @@
+const {makeImmutable} = require('../../../common/state/immutableUtil')
+const platformUtil = require('../../../common/lib/platformUtil')
+const {braveShieldsEnabled} = require('../../../common/state/shieldState')
+const tabUtil = require('../../lib/tabUtil')
+const eventUtil = require('../../../../js/lib/eventUtil')
+const {isNavigatableAboutPage, getBaseUrl} = require('./../../../../js/lib/appUrlUtil')
+const frameStateUtil = require('../../../../js/state/frameStateUtil')
+const siteSettings = require('../../../../js/state/siteSettings')
+const siteUtil = require('../../../../js/state/siteUtil')
+const cx = require('../../../../js/lib/classSet')
+
+// Constants
+const settings = require('../../../../js/constants/settings')
+const appConfig = require('../../../../js/constants/appConfig')
+
+class Navigator extends ImmutableComponent {
+  constructor () {
+    super()
+    this.onBack = this.onBack.bind(this)
+    this.onForward = this.onForward.bind(this)
+    this.onBackLongPress = this.onBackLongPress.bind(this)
+    this.onForwardLongPress = this.onForwardLongPress.bind(this)
+    this.onDoubleClick = this.onDoubleClick.bind(this)
+    this.onDragOver = this.onDragOver.bind(this)
+    this.onDrop = this.onDrop.bind(this)
+    this.onBraveMenu = this.onBraveMenu.bind(this)
+  }
+
+  onNav (e, navCheckProp, navType, navAction) {
+    const activeFrame = frameStateUtil.getActiveFrame(this.props.windowState)
+    const activeTabId = tabUtil.activeTabId(this.props.windowState)
+    const activeTab = activeFrame ? this.props.appState.get('tabs').find((tab) => tab.get('tabId') === activeTabId) : null
+    const isNavigable = isNavigatableAboutPage(getBaseUrl(activeFrame.get('location')))
+    if (e && eventUtil.isForSecondaryAction(e) && isNavigable) {
+      if (activeTab && activeTab.get(navCheckProp)) {
+        appActions.tabCloned(activeTabId, {
+          [navType]: true,
+          active: !!e.shiftKey
+        })
+      }
+    } else {
+      navAction.call(this.activeFrame)
+    }
+  }
+
+  getTotalBlocks (frames) {
+    if (!frames) {
+      return false
+    }
+
+    frames = makeImmutable(frames)
+
+    const ads = frames.getIn(['adblock', 'blocked'])
+    const trackers = frames.getIn(['trackingProtection', 'blocked'])
+    const scripts = frames.getIn(['noScript', 'blocked'])
+    const fingerprint = frames.getIn(['fingerprintingProtection', 'blocked'])
+    const blocked = (ads && ads.size ? ads.size : 0) +
+      (trackers && trackers.size ? trackers.size : 0) +
+      (scripts && scripts.size ? scripts.size : 0) +
+      (fingerprint && fingerprint.size ? fingerprint.size : 0)
+
+    return (blocked === 0)
+      ? false
+      : ((blocked > 99)
+        ? '99+'
+        : blocked)
+  }
+
+  get extensionButtons () {
+    const activeTabId = tabUtil.activeTabId(this.props.windowState)
+    const enabledExtensions = extensionState.getEnabledExtensions(this.props.appState)
```

Comment:
comments about this.props.windowState also apply to anything using this.props.appState